File: src/android/src/com/che/zap/crypto/LocalKeyProvider.kt
package com.che.zap.crypto

import java.security.KeyPair
import java.security.KeyPairGenerator
import java.security.interfaces.ECPrivateKey
import java.security.interfaces.ECPublicKey
import java.security.spec.ECGenParameterSpec

class LocalKeyProvider {

    private lateinit var pair: KeyPair

    fun getPublicKeyBytes(): ByteArray {
        return EncryptionUtils.publicKeyToByteArray(pair.public as ECPublicKey)
    }

    fun getPublicKey(): ECPublicKey {
        return pair.public as ECPublicKey
    }

    fun getPrivateKey(): ECPrivateKey {
        return pair.private as ECPrivateKey
    }

    private fun generateKeyPair() {
        val keyPairGenerator = KeyPairGenerator.getInstance("EC")
        keyPairGenerator.initialize(ECGenParameterSpec("secp256r1"))
        pair = keyPairGenerator.genKeyPair()
    }

    init {
        generateKeyPair()
    }
}

File: src/ios/zwift_play/localKeyProvider.swift
import Foundation
import CryptoKit

class LocalKeyProvider {
    private var keyPair: P256.KeyAgreement.PrivateKey?

    init() {
        generateKeyPair()
    }

    func getPublicKeyBytes() -> Data {
        guard let publicKey = keyPair?.publicKey else {
            fatalError("Key pair not initialized correctly.")
        }
        return publicKey.rawRepresentation
    }

    func getPublicKey() -> P256.KeyAgreement.PublicKey {
        guard let publicKey = keyPair?.publicKey else {
            fatalError("Key pair not initialized correctly.")
        }
        return publicKey
    }

    func getPrivateKey() -> P256.KeyAgreement.PrivateKey {
        guard let keyPair = keyPair else {
            fatalError("Key pair not initialized correctly.")
        }
        return keyPair
    }

    private func generateKeyPair() {
        keyPair = P256.KeyAgreement.PrivateKey()
    }
}
